redis_json/src/ivalue_manager.rs

```diff
@@ -36,13 +37,83 @@ pub struct IValueKeyHolderWrite<'a> {
-            _ => None,
-        })
+#[derive(Debug)]
+pub enum PathValue<'a, 'b: 'a> {
+    /// Mutable reference to an IValue
+    IValue(&'a mut IValue),
+    /// Mutable reference to the IValue array which contains i8s, and its index
+    I8(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains u8s, and its index
+    U8(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains i16s, and its index
+    I16(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains u16s, and its index
+    U16(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains f16s, and its index
+    F16(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains bf16s, and its index
+    BF16(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains i32s, and its index
+    I32(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains u32s, and its index
+    U32(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains f32s, and its index
+    F32(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains i64s, and its index
+    I64(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains u64s, and its index
+    U64(&'b mut IArray, usize),
+    /// Mutable reference to the IValue array which contains f64s, and its index
+    F64(&'b mut IArray, usize),
+}
+
+impl<'a, 'b: 'a> PathValue<'a, 'b> {
+    fn get_from_array(array: &'b mut IArray, index: usize) -> Option<Self> {
+        if index >= array.len() {
+            return None;
+        }
+        let type_tag = array.as_slice().type_tag();
+        let is_heterogeneous = matches!(array.as_mut_slice(), ArraySliceMut::Heterogeneous(_));
+        if is_heterogeneous {
+            return Some(PathValue::IValue(array.get_mut(index).unwrap()));
+        } else {
+            match type_tag {
+                ArrayTag::I8 => Some(PathValue::I8(array, index)),
+                ArrayTag::U8 => Some(PathValue::U8(array, index)),
+                ArrayTag::I16 => Some(PathValue::I16(array, index)),
+                ArrayTag::U16 => Some(PathValue::U16(array, index)),
+                ArrayTag::F16 => Some(PathValue::F16(array, index)),
+                ArrayTag::BF16 => Some(PathValue::BF16(array, index)),
+                ArrayTag::I32 => Some(PathValue::I32(array, index)),
+                ArrayTag::U32 => Some(PathValue::U32(array, index)),
+                ArrayTag::F32 => Some(PathValue::F32(array, index)),
+                ArrayTag::I64 => Some(PathValue::I64(array, index)),
+                ArrayTag::U64 => Some(PathValue::U64(array, index)),
+                ArrayTag::F64 => Some(PathValue::F64(array, index)),
+                ArrayTag::Heterogeneous => unreachable!(),
```

Comment:
This unreachable should not be here. this `match` is itself capable of entirely replacing the outer `if`-`else`